### PR TITLE
[Fix](Nereids)fix type coercion for binary arithmetic

### PR DESCRIPTION
**Pull request by sohardforaname**
# Proposed changes

Issue Number: close #xxx

## Problem summary

support sql like: select true + 1 + '2.0' and prevent select true + 1 + 'x';

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.72 seconds
 load time: 629 seconds
 storage size: 17123573208 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221219154438_clickbench_pr_65275.html

**Comment by morrySnow** [comment]
rebase and fix P0 failed case please~

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/TypeCoercionUtils.java [hunk 1]
```diff
@@ -196,6 +194,8 @@ public static Optional<DataType> findTightestCommonType(DataType left, DataType
             tightestCommonType = right;
         } else if (right instanceof NullType) {
             tightestCommonType = left;
+        } else if (left instanceof BooleanType || right instanceof BooleanType) {
+            tightestCommonType = DoubleType.INSTANCE;
         } else if (left instanceof IntegralType && right instanceof DecimalV2Type
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/util/TypeCoercionUtils.java:199]: this is not right, think about boolean with decimal or with stringliketype

**Comment by morrySnow** [comment]
current master could pass all new P0 case, but `select true + 1 + 'x';` is not same with legacy planner.
```
mysql> set enable_nereids_planner=true;
mysql> select true + 1 + 'x';
+--------------------+
| ((TRUE + 1) + 'x') |
+--------------------+
|               NULL |
+--------------------+
1 row in set (0.01 sec)

mysql> set enable_nereids_planner=false;
Query OK, 0 rows affected (0.00 sec)

mysql> select true + 1 + 'x';
ERROR 5012 (HY000): errCode = 2, detailMessage = 'x' is not a number
```

**Review by morrySnow — APPROVED** [review]

**Pull request merged**